### DISCUSSION TITLE
How to be able to use WMI with 'root\default' namespace? Is this a limitation of the framework?

### DISCUSSION THREAD

**Discussion by bogdan-patraucean** [microsoft/WindowsAppSDK#4286] [Q&A] [answered]
```
ManagementScope scope = new (@"root\DEFAULT");
ManagementPath path = new ("SystemRestore");
ManagementClass mClass = new (scope ,path , new ObjectGetOptions());
var restorePoints = mClass.GetInstances().Count;
```

This piece of code works perfectly in a console app, but fails on a WinAppSdk project (both packaged and unpackaged) with the following exception: `System.Runtime.InteropServices.COMException (0x80040154)`

Some relevant info:
- app targets 64 bits architecture
- app is configured to run elevated
- WMI is enabled
- classes from other namespaces work like a charm

Here is the sample in case you want to debug or something: [System Restore points.zip](https://github.com/microsoft/WindowsAppSDK/files/14733779/System.Restore.points.zip)

**Comment by ghost1372** [comment]
@bogdan-patraucean 

It seems that this code does not have the access

![image](https://github.com/microsoft/WindowsAppSDK/assets/9213496/8279fc6a-be61-4008-a659-4c91b5b5bded)

but, you can use elevated mode (Run as Admin) which works fine

![image](https://github.com/microsoft/WindowsAppSDK/assets/9213496/b8fa1bb4-bd79-4e8b-a317-48cfc7f69a70)

**Reply by bogdan-patraucean** [reply]
For me it doesn't work even elevated. Did you do something different compared to my sample?

**Reply by ghost1372** [reply]
> For me it doesn't work even elevated. Did you do something different compared to my sample?

I just installed System.Management nuget package
Wasdk v1.5
Dotnet 8
And codes you see in images

**Reply by bogdan-patraucean** [reply]
What type of project did you create?

**Reply by ghost1372** [reply]
Hi @bogdan-patraucean 
i tested your sample and it seems that this is a WinUI 3 Template (2 Project) issue, i am using Single MSIX Project.
you can download my sample here, in UnPackaged Mode you can run exe as Admin or, just run your visual studio in Admin mode and run project
[App9.zip](https://github.com/microsoft/WindowsAppSDK/files/14735220/App9.zip)

**Reply by bogdan-patraucean** [reply]
@ghost1372, yeah, it seems to be a problem only when packaged.

**Comment by DarranRowe** [comment]
HRESULT 0x80040154 is REGDB_E_CLASSNOTREG.
For various reasons, are you sure that this is the right error code?

--Edit--
Hmm, this seems related to running it as packaged. Unpackaged works fine if you remember to bootstrap the WAS.

![Screenshot 2024-03-23 232555](https://github.com/microsoft/WindowsAppSDK/assets/52577874/0701f717-f6b5-4d4b-bbed-b05118eaae64)

**Reply by bogdan-patraucean** [reply]
How did you do that bootstrap thing? Also, any idea why it works as unpackaged (with bootstrap) but not packaged?

**Reply by DarranRowe** [reply]
The easiest way is to add:

```
    <WindowsPackageType>None</WindowsPackageType>
    <WindowsAppSDKBootstrapAutoInitializeOptions_OnPackageIdentity_NoOp>true</WindowsAppSDKBootstrapAutoInitializeOptions_OnPackageIdentity_NoOp>
```

WindowsPackageType tells the build process to compile in the auto initialiser for the Windows App Runtime. You can see the MddBootstrapAutoInitializer.cs added to the project. This automatically happens for C# projects when you add WindowsPackageType.
The other one is to tell the bootstrapper to not result in an error if the bootstrapper runs when the application uses it when the application is packaged. This is a very simple way to make the application dual mode, it will automatically bootstrap when unpackaged, but then works when it is packaged where the bootstrapper is unneeded.

![Screenshot 2024-03-24 001433](https://github.com/microsoft/WindowsAppSDK/assets/52577874/89d8c184-86cc-483c-8f7a-495902261cfc)

Although this has issues with the single project WinUI 3 projects. If you can't use WindowsPackageType, then you would have to use the [bootstrapper API](https://learn.microsoft.com/en-us/windows/apps/api-reference/cs-bootstrapper-apis/) manually. The bootstrapping is required to use the Windows App SDK in an unpackaged application.

As to why it doesn't work. No idea, considering where the error seems to be coming from, I would guess that the System.Management library doesn't work with packaged applicaitons.

**Reply by bogdan-patraucean** [reply]
Thanks a lot!
Other classes are working well with a packaged type, only this SystemRestore one I found to be problematic.
On the other hand, I tried to bootstrap with your instructions but still doesn't work, can you please send me your working sample so I can compare?

**Reply by DarranRowe** [reply]
I simply took your posted sample in your project that you provided above and added those two settings to the .csproj.
However, the important thing is that you must change the start up project to the unpackaged project. For your sample, you MUST change it from App2 (Package) to App2.
Although give me a few minutes and I will modify your sample into a way that will work.

--EDIT--

[System.Restore.points.zip](https://github.com/microsoft/WindowsAppSDK/files/14733993/System.Restore.points.zip)

There, but as I wrote, it is CRITICAL that you change the start up project. You can do this in the menu region by changing the project to App2

![Screenshot 2024-03-24 004801](https://github.com/microsoft/WindowsAppSDK/assets/52577874/2c09904c-447e-4f93-8728-d4e76ac02c40)

You can also verify that the correct project is set to the start up project by looking in Solution Explorer and verifying that App2 is the project that is in bold.

![Screenshot 2024-03-24 004721](https://github.com/microsoft/WindowsAppSDK/assets/52577874/f96ba124-fabd-47e0-aa42-e7fefa31d7f5)

This means that it will execute App2.exe directly and not try to deploy it as a package.

Also, I may have change the project to use .NET 7. If it is .NET 7 and not 6 like your project originally had then you can just change it back. The .NET version shouldn't have any impact. I just don't have a .NET 6 SDK installed on my system and the WinUI templates default to .NET 6.

**Reply by bogdan-patraucean** [reply]
I tried, even with your sample, but it crashes on opening. Or it just doesn't open, I even tried to execute the exe directly from the bin folder. I'll try on another computer. But anyway, should I open an issue for this? If other classes work perfectly besides the `SystemRestore` in a packaged project?

Edit: it works on another computer the way you described. But the code is not the problem, it does the same for an empty app.

**Reply by DarranRowe** [reply, marked as answer]
Well, I took the Windows App SDK completely out of the equation and used the code in your original post in a simple console application. It only referenced the System.Management package. I packaged this up using a C# application packaging project and it fails with the exact same error.

![Screenshot 2024-03-24 100712](https://github.com/microsoft/WindowsAppSDK/assets/52577874/87abff18-e814-43e1-8161-0640753cac24)

This was using:
```
using System.Management;

namespace ConsoleApp1
{
    internal class Program
    {
        static void Main(string[] args)
        {
            ManagementScope scope = new(@"root\DEFAULT");
            ManagementPath path = new("SystemRestore");
            ManagementClass mClass = new(scope, path, new ObjectGetOptions());
            var restorePoints = mClass.GetInstances().Count;

            Console.WriteLine(restorePoints);
        }
    }
}
```

So your best bet is to ask about this on the [dotnet/runtime](https://github.com/dotnet/runtime) repository.

**Reply by bogdan-patraucean** [reply]
@DarranRowe, thanks a lot for your help and your time, I'll open another discussion in that repo, maybe we can fix this. Can you point me to the resources that would help me make what you did? packaging a console app? I never did this. I think it would help if I open the discussion with a repro sample.

**Reply by DarranRowe** [reply]
You can use [this](https://learn.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-packaging-dot-net) as a guide. It basically creates the same kind of packaging project that the WinUI 3 uses when you create one of those with a separate packaging project.
